go/cmd/dolt/commands/checkout.go

```diff
@@ -82,276 +87,137 @@ func (cmd CheckoutCmd) Exec(ctx context.Context, commandStr string, args []strin
+		return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usagePrt)
 	}
+	defer closeFunc()
 
 	branchOrTrack := apr.Contains(cli.CheckoutCoBranch) || apr.Contains(cli.TrackFlag)
 	if (branchOrTrack && apr.NArg() > 1) || (!branchOrTrack && apr.NArg() == 0) {
 		usagePrt()
 		return 1
 	}
 
-	if branchOrTrack {
-		verr := checkoutNewBranch(ctx, dEnv, apr)
-		return HandleVErrAndExitCode(verr, usagePrt)
-	}
-
-	name := apr.Arg(0)
-	force := apr.Contains(cli.ForceFlag)
-
-	if len(name) == 0 {
-		verr := errhand.BuildDError("error: cannot checkout empty string").Build()
-		return HandleVErrAndExitCode(verr, usagePrt)
-	}
-
-	if isBranch, err := actions.IsBranch(ctx, dEnv.DoltDB, name); err != nil {
-		verr := errhand.BuildDError("error: unable to determine type of checkout").AddCause(err).Build()
-		return HandleVErrAndExitCode(verr, usagePrt)
-	} else if isBranch {
-		verr := checkoutBranch(ctx, dEnv, name, force)
-		return HandleVErrAndExitCode(verr, usagePrt)
-	}
-
-	// Check if the user executed `dolt checkout .`
-	if apr.NArg() == 1 && name == "." {
-		roots, err := dEnv.Roots(ctx)
-		if err != nil {
-			return HandleVErrAndExitCode(errhand.BuildDError(err.Error()).Build(), usagePrt)
-		}
-		headRef, err := dEnv.RepoStateReader().CWBHeadRef()
-		if err != nil {
-			return HandleVErrAndExitCode(errhand.BuildDError(err.Error()).Build(), nil)
+	var branchName string
+	if apr.Contains(cli.CheckoutCoBranch) {
+		branchName, _ = apr.GetValue(cli.CheckoutCoBranch)
+	} else if apr.Contains(cli.TrackFlag) {
+		if apr.NArg() > 0 {
+			usagePrt()
+			return 1
 		}
-		ws, err := dEnv.WorkingSet(ctx)
-		if err != nil {
-			HandleVErrAndExitCode(errhand.BuildDError(err.Error()).Build(), usagePrt)
-		}
-		verr := actions.ResetHard(ctx, dEnv, "HEAD", roots, headRef, ws)
-		return handleResetError(verr, usagePrt)
+		remoteAndBranchName, _ := apr.GetValue(cli.TrackFlag)
+		_, branchName = actions.ParseRemoteBranchName(remoteAndBranchName)
+	} else if apr.NArg() > 0 {
+		branchName = apr.Arg(0)
 	}
 
-	verr := checkoutTables(ctx, dEnv, args)
-	if verr != nil && apr.NArg() == 1 {
-		verr = checkoutRemoteBranchOrSuggestNew(ctx, dEnv, name)
+	sqlQuery, err := generateCheckoutSql(args)
+	if err != nil {
+		return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usagePrt)
 	}
 
-	return HandleVErrAndExitCode(verr, usagePrt)
-}
-
-func checkoutNewBranch(ctx context.Context, dEnv *env.DoltEnv, apr *argparser.ArgParseResults) errhand.VerboseError {
-	var newBranchName string
-	var remoteName string
-	var remoteBranchName string
-	var startPt = "head"
+	rows, err := GetRowsForSql(queryEngine, sqlCtx, sqlQuery)
 
-	if apr.NArg() == 1 {
-		startPt = apr.Arg(0)
+	if err != nil {
+		// In fringe cases the server can't start because the default branch doesn't exist, `dolt checkout <existing branch>`
+		// offers an escape hatch.
+		if dEnv != nil && !branchOrTrack && strings.Contains(err.Error(), "cannot resolve default branch head for database") {
```

Comment:
This error occurs even when the queryist is a SQLEngine. It's a necessary escape hatch for the case where the checked-out branch is deleted.

I removed the dEnv nil check though, since as you pointed out it's always set.